### DISCUSSION TITLE
Creating the docker for open3d-ml

### DISCUSSION THREAD

**Discussion by shayannikoohemat** [isl-org/Open3D#5849] [Q&A]
I am trying to create a **docker image for Open3d-ML** on an Ubuntu20.04 base image following this [link](http://www.open3d.org/docs/release/docker.html). After creating the docker image and running it I get this error: 
`raise Exception('Open3D was not built with PyTorch support!')
`

**Here is the Docker image:**
```
FROM ubuntu:20.04

# Install Open3D system dependencies and pip
RUN apt-get update && apt-get install --no-install-recommends -y \
    libgl1 \
    libgomp1 \
    python3-pip \
    && rm -rf /var/lib/apt/lists/*

# Install Open3D from the PyPI repositories
RUN python3 -m pip install --no-cache-dir --upgrade pip && \
   python3 -m pip install --no-cache-dir --upgrade open3d


## install PyTorch for open3d-ml
#RUN pip install torch==1.12.0+cpu torchvision==0.13.0+cpu torchaudio==0.12.0 --extra-index-url https://download.pytorch.org/whl/cpu
RUN pip install torch torchvision torchaudio --extra-index-url https://download.pytorch.org/whl/cpu

RUN pip install packaging
```

**Build the docker:**
`docker build -t open3d-ml-ubuntu20:latest Dockerfile .
`

**Run the docker:**
`docker run -it --rm --name o3d-ubuntu20-1 open3d-ml-ubuntu20
`

Then inside the docker:
```
root@2131cc1d0aa6:/# python3 --version
Python 3.8.10
root@2131cc1d0aa6:/# pip freeze | grep open3d
open3d==0.16.0
root@0fd2a511eb1f:/# pip3 freeze | grep torch
torch==1.13.1
torchaudio==0.13.1
torchvision==0.14.1
```

### Error Message
Importing open3d works fine with no error but when I import `open3d.ml.torch` I get the error:
```
root@2131cc1d0aa6:/# python3
Python 3.8.10 (default, Nov 14 2022, 12:59:47) 
[GCC 9.4.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import open3d.ml.torch
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/local/lib/python3.8/dist-packages/open3d/ml/torch/__init__.py", line 34, in <module>
    raise Exception('Open3D was not built with PyTorch support!')
Exception: Open3D was not built with PyTorch support!
>>> 
```

### Additional information
I tried also below options:
* I tried with `Ubuntu22.04` and `python 3.10` but the same error. 
* I tried to install Pytorch with the versions specified in the [requirements-torch.txt](https://github.com/isl-org/Open3D-ML/blob/master/requirements-torch.txt) with this command in the docker:
`RUN pip install torch==1.12.0+cpu torchvision==0.13.0+cpu torchaudio==0.12.0 --extra-index-url https://download.pytorch.org/whl/cpu`
Error:
```
#7 1.027 ERROR: Could not find a version that satisfies the requirement torch==1.12.0+cpu (from versions: 1.8.0, 1.8.1, 1.9.0, 1.10.0, 1.10.1, 1.10.2, 1.11.0, 1.12.0, 1.12.1, 1.13.0, 1.13.1)
#7 1.027 ERROR: No matching distribution found for torch==1.12.0+cpu
```
* I tried a specific version of open3d e.g. `open3d==0.12.1` but I get the error: `ERROR: No matching distribution found for open3d`

### Related disucssions
[5123](https://github.com/isl-org/Open3D/discussions/5123) 

**System Info** 
(but note I'm building a docker with Ubuntu=20.04 base image)
- MacBook Pro
- macOS 12.6
- Kernel Version: Darwin 21.6.0

Any help on how to build a docker for Open3d-ML is appreciated.

**Comment by shayan-nikoo** [comment]
I noticed the Docker by default runs with the native OS, so if I want to run a docker with Linux base image on an ARM64 (Mac OS) I should pass the tag `--platform=linux/amd64` during `build` and `run`.

The below Docker works fine on a Mac Os M1 system. 
```
# works on MacOS and linux
ARG PLATFORM=linux/amd64
ARG PYTHON_DIST=3.10.9-slim-bullseye

FROM --platform=${PLATFORM} python:${PYTHON_DIST}

ENV DEBIAN_FRONTEND="noninteractive"

# Install Open3D system dependencies
RUN apt-get update && apt-get install --no-install-recommends -y \
    libgl1 \
    libgomp1 \
    && rm -rf /var/lib/apt/lists/*

# Install Open3D from the PyPI repositories
RUN python3 -m pip install --no-cache-dir --upgrade pip && \
   python3 -m pip install --no-cache-dir --upgrade open3d==0.16.0 tensorboard==2.11.2

# install pytorch
RUN pip install \
  torch==1.12.0+cpu \
  torchvision==0.13.0+cpu \
  torchaudio==0.12.0+cpu \
  --extra-index-url https://download.pytorch.org/whl/cpu \
  && rm -rf /root/.cache/pip
```
Note `--platform=${PLATFORM} ` at the `FROM` line of the docker.

Then build and run it with:
```
docker build -t open3dml016-pytoch1120-cpu-py310:latest --platform=linux/amd64 Dockerfile . 

docker run -it --rm --name o3d-troch-cpu open3dml016-pytoch1120-cpu-py310:latest
```
